### PR TITLE
[FLINK-25024][docs] Add Changelog backend docs

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Add documentation for State Changelog (FLIP-158):
 - state-backends
 - configuration
 - metrics

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes (internal: flink-docs -> flink-dstl-dfs)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? see above

### CODE REVIEW

**Comment by rkhachatryan** [comment]
Updated `Installation` section after agreeing on including all jars into the main Flink jar (#18432).
Added `Limitations` and `Upgrading` sections,
mention of Recovery time.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [inline docs/content/docs/ops/state/state_backends.md:352]: 1. Add some context here:
Current Incremental Checkpoints depend on the implementation of different types of state backends. For example, for rocksdb, compaction happens when ... 

2. Explain a bit why compaction is bad
Compaction may cause more data to be uploaded, and more time to upload....

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [inline docs/content/docs/ops/state/state_backends.md:355]: "is checkpointed" => "is snapshotted"

It may confuse people with the normal Flink checkpointing procedure.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [inline docs/content/docs/ops/state/state_backends.md:358]: "as well as synchronous phase (in particular, long-tail)"

Could you explain a bit more. Can not infer directly from what stated above.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [inline docs/content/docs/ops/state/state_backends.md:363]: upload state change/changelogs

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [inline docs/content/docs/ops/state/state_backends.md:366]: mark here to remove todo in the final version.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [inline docs/content/docs/ops/state/state_backends.md:371]: `However, recovery time combined with checkpoint
duration will likely be still lower than in non-changelog setup, providing lower end-to-end latency even in failover
case.`

I am not sure of that.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [inline docs/content/docs/ops/state/state_backends.md:441]: Include these in the limitation section?

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:330]: Yes, that's what I was going to do, I just wanted to get the approval first for the English version (otherwise translation would be inconsistent).

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:347]: I agree, it's confusing.

What I mean here is minimizing the synchronous phase duration by using some form of persistent data structures (CoW for Heap and immutable SSTables in RocksDB). 
This is [called](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/ops/state/state_backends/#the-embeddedrocksdbstatebackend) asynchronous checkpoints/snapshots.


WDYT about
```
2. Snapshot creation time (so-called synchronous phase), reduced by immutable (or asynchronous) snapshots. 
```
Or maybe just remove the clarification?

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:352]: 1. I think explaining when the compaction can be triggered is outside of scope of this section. There are several compaction algorithms, to start with.
2. Agree

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:371]: Could you elaborate a bit?

In the experiments, we saw a reduction of checkpoint duration by minutes, and increase of recovery times by tens of seconds. This is not guaranteed of course, but I think it is likely.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [reply docs/content/docs/ops/state/state_backends.md:347]: That's what I am not sure about.

Snapshot creation time: (isn't that just preparing for changelogs to be uploaded)? 
And previously snapshot creat time: flush + prepare SSTs to be uploaded?

I think most of the time reduced is from the async phase (uploading before ahead, so the async phase does not take very long time).

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [reply docs/content/docs/ops/state/state_backends.md]: I am not asking to explain the entire different compaction algorithms, but since you mentioned "at least one task in every checkpoint that uploads a lot of data (e.g. after compaction)." You need to explain why.

That's what I mean by providing some context: what causes to upload a lot data; that's compaction, then why compaction causes more data to upload... e.t.c

I do not think people can infer directly why compaction can cause more data to upload until you explain at least a little bit of different level of compaction e.t.c.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [reply docs/content/docs/ops/state/state_backends.md:373]: Recovery time really depends on how much data to restore.

I also saw a lot of cases that recovery time increased by tens of seconds but checkpoint duration does not decrease that much.

What we showed in the graph is a job instance with 196GB state size, and there are a lot of cases that the above statement does not hold.

I do not think we should mix recovery time with checkpoint duration here.

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:347]: > Snapshot creation time: (isn't that just preparing for changelogs to be uploaded)?
> And previously snapshot creat time: flush + prepare SSTs to be uploaded?

Yes.

> Would these two parts significantly different?
> I think most of the time reduced is from the async phase

For changelog, it's usually under 100ms, without it can take tens of seconds (above p99%). That can sound not much, but this phase blocks the entire processing. I describe it [here](https://github.com/apache/flink/pull/18431#discussion_r790436952).

But actually this statement is about reduction by using Async checkpoints which were implemented long time ago, not related to the Changelog.

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md]: Explained, PTAL.

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md]: I've expanded this section to address (2), do you still have any concerns about (1)?

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:373]: > I also saw a lot of cases that recovery time increased by tens of seconds but checkpoint duration does not decrease that much.

If checkpoint duration doesn't decrease significantly then it's probably not a suitable case to use Changelog. So the question I think is what's the ratio between these times and how to put it in the docs.

> I do not think we should mix recovery time with checkpoint duration here.

Reducing **effective** recovery time is actually one of the goals of the project (please see FLIP [Motivation](https://cwiki.apache.org/confluence/display/FLINK/FLIP-158%3A+Generalized+incremental+checkpoints#FLIP158:Generalizedincrementalcheckpoints-Motivation) and discussions); and the mean to achieve this is by having less data to replay for the whole pipeline (while having data to replay by Changelog on Task level).

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:347]:
```suggestion
2. Snapshot creation time (so-called synchronous phase), addressed by asynchronous snapshots
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:353]:
```suggestion
Upload time can be decreased by [incremental checkpoints]({{< ref "#incremental-checkpoints" >}}).
However, most incremental state backends perform some form of compaction periodically, which results in re-uploading the
old state in addition to the new changes. In large deployments, the probability of at least one task uploading lots of
data tends to be very high in every checkpoint.
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:357]:
```suggestion
With Changelog enabled, Flink uploads state changes continuously and forms a changelog. On checkpoint, only the relevant
part of this changelog needs to be uploaded. The configured state backend is snapshotted in the
background periodically. Upon successful upload, the changelog is truncated.
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:362]:
```suggestion
However, resource usage is higher:
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:373]:
```suggestion
Recovery time is another thing to consider. Depending on the `state.backend.changelog.periodic-materialize.interval` setting,
the changelog can become lengthy and replaying it may take more time. However, recovery time combined with checkpoint
duration will likely still be lower than in non-changelog setups, providing lower end-to-end latency even in failover
case.
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:381]:
```suggestion
Changelog JARs are included into the standard Flink distribution.

Make sure to [add]({{< ref "docs/deployment/filesystems/overview" >}}) the necessary filesystem plugins.
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:385]:
```suggestion
Here is an example configuration in YAML:
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:398]:
```suggestion
Please refer to the [configuration section]({{< ref "docs/deployment/config#state-changelog-options" >}}) for other options.
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:400]:
```suggestion
Changelog can also be enabled or disabled per job programmatically:
```

File: docs/content/docs/ops/state/state_backends.md
Comment (infoverload) [inline docs/content/docs/ops/state/state_backends.md:426]:
```suggestion
If a task is backpressured by writing state changes, it will be shown as busy (red) in the UI.
```

**Review by infoverload — COMMENTED** [review]
I gave a preliminary review

**Comment by rkhachatryan** [comment]
Thanks a lot @infoverload, I've accepted your suggestions.
FYI @curcur, the PR was updated.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [reply docs/content/docs/ops/state/state_backends.md:373]: What I want to say is there is no evidence showing that recovery time is guaranteed to be reduced. 
1. "less data replay" has an assumption that checkpoint duration is reduced (there is not always a significant reduction as I mentioned above).
2. Replaying of changelog to DB is an additional cost for recovery with changelog vs. normal recovery. 

1+2 made me not be convinced with 
"However, recovery time combined with checkpoint duration will likely be still lower than in non-changelog setup, providing lower end-to-end latency even in failover case."

**Comment by curcur** [comment]
> Thanks a lot @infoverload, I've accepted your suggestions. FYI @curcur, the PR was updated.

thanks a lot @rkhachatryan and @infoverload , I will take a look!

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [inline docs/content/docs/ops/state/state_backends.md:347]: `asynchronous snapshots`

What are asynchronous snapshots? and how it is different from `asynchronous phase`

Do you mean the materialization is separated and independent with the checkpointing/snapshotting?

**Comment by curcur** [comment]
Hey @rkhachatryan , I've read through he doc again

1. Two things for doc itself: one is about async snapshot; the other is about recovery time (details are put inline)
2. For the commits, is there any way to squash the commit to one? Now it includes 13 commits (not counting fix-up)...
3. Should we mention leftovers in the limitation? (This is not a must but I feel we should be honest on this part)

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:347]: > What are asynchronous snapshots? and how it is different from asynchronous phase

synchronous phase = create a snapshot
asynchronous phase = upload a snapshot

Implemented naively, sync phase would duplicate **all** the state (copying all in-memory tables and/or on-disk files). So that async phase and processing can proceed concurrently.

With Asynchronous snapshots (probably a misnormer), sync phase only flushes or modifies small portions of data, as I explained [above](https://github.com/apache/flink/pull/18431#discussion_r790783796).

> Do you mean the materialization is separated and independent with the checkpointing/snapshotting?

No. As I said [above](https://github.com/apache/flink/pull/18431#discussion_r791455950), asynchronous snapshots are not related to Changelog.

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:373]: There is **no guarantee**, but there **is evidence** for some cases. 
And this is also true for the checkpoint duration (there is even guarantee that it will increase in some cases, e.g. many updates of a single key).

That's why I put it as "likely". Do you prefer some other wording?

Or I could add something like "However, it's also possible that the effective recovery time will increase, depending on the actual ratio of the aformentioned times", WDYT?

Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md]: I've added the latter statement in 2512d75, PTAL.

File: docs/content/docs/ops/state/state_backends.md
Comment (curcur) [reply docs/content/docs/ops/state/state_backends.md:347]: **Without Changelog:**
Sync phase: flush in-mem data -> disk, prepare what-ssts to upload

**With Changelog:**
Sync phase: prepare state-changes to upload

From the differences above, I can not infer what async snapshots refer to and can not see what problem the "async snapshots are addressed"? Or if I misunderstand the above process?

Comment (curcur) [reply docs/content/docs/ops/state/state_backends.md:347]: OK, I finally got what you said.

Is there any reference you can use here? similar to Unaligned checkpoints and Buffer debloating?

It would be too confusing otherwise.

**Comment by curcur** [comment]
I've accept the PR. Thanks for the effort Roman!

**Review by curcur — APPROVED** [review]

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md:347]: Sure, I'll linkt to [RocksDB](https://nightlies.apache.org/flink/flink-docs-master/docs/ops/state/state_backends/#the-embeddedrocksdbstatebackend) section where they are mentioned.

File: docs/content/docs/ops/state/state_backends.md
Comment (rkhachatryan) [reply docs/content/docs/ops/state/state_backends.md]: FLINK-25867

**Comment by rkhachatryan** [comment]
Thanks for the review @curcur and @infoverload .
I'll merge the PR once the build is green.

**Pull request merged**
